architecture/ambient/lifecycle.md

```diff
@@ -0,0 +1,136 @@
+This can help ensure that clients do not hold onto pooled connections for too long.
+This is not strictly required; Ztunnel clients have both keepalives (which will start to fail once the other end is shut down) and pool idle timeouts (which will close the connection when it has had no activity for some period; because the backend is gone, there will be no activity).
+However, it would be ideal to give immediate feedback rather than relying on these mechanisms.
+
+### Pod Shutdown Implementation
+
+[ZDS](../../pkg/zdsapi/zds.proto) exposes a `DelWorkload` message to signal to Ztunnel that a pod has been terminated.
+This can be triggered either by a Pod being actually deleted, or has transitioned to a [terminal phase](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/).
+At this point, the pod and all containers have fully shut down, and we can tear everything down.
+
+Because the application(s) are already entirely gone, there is no need to have any "draining" time -- we can immediately terminate.
+Furthermore, because the Pod is torn down, we cannot even send any data at all (at least on most clusters, the `veth` is torn down upon Pod deletion).
+Notably, this prevents us from sending a GOAWAY to notify peers we have shutdown.
+
+As a result, while we meet our two must-haves (send and accept traffic throughout the entire pod lifetime), we do not have an optimal solution here yet.
+Perhaps hooking into the [CNI `DEL`](https://github.com/containernetworking/cni/blob/main/SPEC.md#del-remove-container-from-network-or-un-apply-modifications) could be a path forward here.
+Alternatively, we could send the GOAWAY prior to the pod termination (assuming this allows existing and new connections to function still).
+
+## Ztunnel Shutdown/Upgrade/Restart
+
+This could be upgrading to a new version, or changing some part of the DaemonSet spec, etc.
+
+### Ztunnel Shutdown Requirements
+
+Our high level requirement is to minimize traffic disruption.
+
+However, because Ztunnel operates only at Layer 4, we are in a tricky position:
+* TCP is stateful, so we cannot really just pass state over to the other process.
+  If we operated at L3 we could just start processing packets in the new Ztunnel.
+* We do not operate at L7, so we cannot signal to the application, in any way, that it should reconnect (to the new Ztunnel).
+
+Realistically, that means the best we can do is:
+* Ensure that, at any point in time, any _new_ connections will succeed. There is no period where new connections are dropped.
+* Provide some period of time for the old Ztunnel to continue processing any established connections.
+
+If this period of time ("drain period") is longer than any connections, there is zero impact on applications at all.
+If not, however, at some point the TCP connection will be torn down.
+Depending on the application, this may have a range of impacts (terminating in-flight transactions, etc).
```

Comment:
```suggestion
Depending on the application, this may have a range of impacts (terminating in-flight transactions, etc). In general, applications should never assume the network is 100% reliable and that no connections should be disrupted, or they will generally have[ stability problems in the context of any distributed system](https://en.wikipedia.org/wiki/Fallacies_of_distributed_computing)
```